xbmc/addons/interfaces/General.cpp

```diff
@@ -0,0 +1,198 @@
+  addonInterface->toKodi->kodi->get_localized_string = get_localized_string;
+  addonInterface->toKodi->kodi->unknown_to_utf8 = unknown_to_utf8;
+  addonInterface->toKodi->kodi->queue_notification = queue_notification;
+  addonInterface->toKodi->kodi->queue_notification_from_type = queue_notification_from_type;
+  addonInterface->toKodi->kodi->queue_notification_with_image = queue_notification_with_image;
+  addonInterface->toKodi->kodi->get_dvd_menu_language = get_dvd_menu_language;
+}
+
+void Interface_General::DeInit(AddonGlobalInterface* addonInterface)
+{
+  if (addonInterface->toKodi && /* <-- Safe check, needed so long old addon way is present */
+      addonInterface->toKodi->kodi)
+  {
+    free(addonInterface->toKodi->kodi);
+    addonInterface->toKodi->kodi = nullptr;
+  }
+}
+
+void Interface_General::open_settings_dialog(void* kodiBase)
+{
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  if (addon == nullptr)
+  {
+    CLog::Log(LOGERROR, "Interface_General::%s - invalid data (addon='%p')", __FUNCTION__, addon);
+    return;
+  }
+
+  // show settings dialog
+  AddonPtr addonInfo;
+  if (CAddonMgr::GetInstance().GetAddon(addon->ID(), addonInfo))
+    CGUIDialogAddonSettings::ShowAndGetInput(addonInfo);
+}
+
+char* Interface_General::get_localized_string(void* kodiBase, long dwCode)
+{
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  if (!addon)
+  {
+    CLog::Log(LOGERROR, "Interface_General::%s - invalid data (addon='%p')", __FUNCTION__, addon);
+    return nullptr;
+  }
+
+  if (g_application.m_bStop)
+    return nullptr;
+
+  std::string string;
+  if ((dwCode >= 30000 && dwCode <= 30999) || (dwCode >= 32000 && dwCode <= 32999))
+    string = g_localizeStrings.GetAddonString(addon->ID(), dwCode).c_str();
+  else
+    string = g_localizeStrings.Get(dwCode).c_str();
+
+  char* buffer = strdup(string.c_str());
+  return buffer;
+}
+
+char* Interface_General::unknown_to_utf8(void* kodiBase, const char* source, bool& ret, bool failOnBadChar)
+{
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  if (!addon || !source)
+  {
+    CLog::Log(LOGERROR, "Interface_General::%s - invalid data (addon='%p', source='%p')", __FUNCTION__, addon, source);
+    return nullptr;
+  }
+   
+  std::string string;
+  ret = g_charsetConverter.unknownToUTF8(source, string, failOnBadChar);
+  char* buffer = strdup(string.c_str());
+  return buffer;
+}
+
+void Interface_General::get_dvd_menu_language(void* kodiBase, char& language, unsigned int& iMaxStringSize)
+{
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  if (!addon)
+  {
+    CLog::Log(LOGERROR, "Interface_General::%s - invalid data (addon='%p')", __FUNCTION__, addon);
+    return;
+  }
+
+  std::string str = g_langInfo.GetDVDMenuLanguage();
+  strncpy(&language, str.c_str(), iMaxStringSize);
+  iMaxStringSize = str.length();
+}
+
+void Interface_General::queue_notification(void* kodiBase, const int type, const char* message)
```

Comment:
Is for a KaiToast Message who have already "void" as return, no need to have a return for me.